aptos-move/aptos-e2e-comparison-testing/src/execution.rs

```diff
@@ -202,152 +233,180 @@ impl Execution {
                     cur_version, res_ok.0, res_ok.1
-                );
+                ));
             } else {
-                println!(
+                self.output_result_str(format!(
                     "execution error {} at version: {}, error",
                     res.unwrap_err(),
                     cur_version
-                );
+                ));
             }
         }
     }
 
     fn execute_code(
         &self,
-        state: &FakeDataStore,
-        features: &Features,
+        version: Version,
+        state: FakeDataStore,
         package_info: &PackageInfo,
         txn: &Transaction,
         compiled_package_cache: &HashMap<PackageInfo, HashMap<ModuleId, Vec<u8>>>,
+        debugger_opt: Option<Arc<dyn AptosValidatorInterface + Send>>,
     ) -> Option<Result<(WriteSet, Vec<ContractEvent>), VMStatus>> {
         let executor = FakeExecutor::no_genesis();
         let mut executor = executor.set_not_parallel();
-        *executor.data_store_mut() = state.clone();
+        *executor.data_store_mut() = state;
         if let Transaction::UserTransaction(signed_trans) = txn {
             let sender = signed_trans.sender();
             let payload = signed_trans.payload();
             if let TransactionPayload::EntryFunction(entry_function) = payload {
-                // always load 0x1 modules
+                // Always load 0x1 modules
                 load_aptos_packages_to_executor(&mut executor, compiled_package_cache);
-                // Load other modules
+                // Load modules
                 if package_info.is_compilable() {
                     load_packages_to_executor(&mut executor, package_info, compiled_package_cache);
                 }
                 let mut senders = vec![sender];
                 senders.extend(TransactionMetadata::new(signed_trans).secondary_signers);
-                return Some(executor.try_exec_entry_with_features(
-                    senders,
-                    entry_function,
-                    features,
-                ));
-            } else if let TransactionPayload::Multisig(multi_sig) = payload {
-                assert!(multi_sig.transaction_payload.is_some());
-                println!("Multisig transaction is not supported yet");
+                if let Some(debugger) = debugger_opt {
+                    let data_view =
+                        DataStateView::new(debugger, version, executor.data_store().clone());
+                    return Some(executor.try_exec_entry_with_resolver(
+                        senders,
+                        entry_function,
+                        &data_view.as_move_resolver(),
+                    ));
+                } else {
+                    return Some(executor.try_exec_entry_with_resolver(
+                        senders,
+                        entry_function,
+                        &executor.data_store().clone().as_move_resolver(),
```

Comment:
Why? try_exec would use executor.data_store() and modify it, so I guess the main problem here us that you want to run something multiple times from the same state (and "uncommit") any modifications?

Reply:
In the current design, each txn has its own state so the corresponding state is only run once without any modifications. In the offline execution mode, the data state is from the file (obtained in the dump mode). In the online execution mode, the data state is from DebuggerStateView.